### ISSUE TITLE
Ticketing - leave BPT

### ISSUE THREAD

**Issue by bethlakshmi** [bethlakshmi/gbe-divio-djangocms-python2.7#206]
The ticket conversation I remember is NOT a GitHub issue, it was an email!

**Scratch**:
OK — you’re right.  The BPT ticket isn’t there for some weird reason.

That said, there is this ticket which seems to tackle part of what I want to tackle.  

The short version is, I want to get us away from BPT handling our cash.  Our options are:

- Leave everything the way it is except use PayPal through BPT as our cash handler.  It has the advantage that we have a system in place that works, and we don’t have to try and port the existing ticket info.
- Move everything to a new ticketing agency (like www.SimpleTix.com).  Has the advantage that we get the money.  Has the disadvantage that Beth has to figure out how to connect our site to theirs, Scratch has to create all of the various ticket items again, and someone has to figure out how to port the existing sales over to the new site.
- Create our own ticketing system.  Even more work for Beth and Scratch.  There do seem to be a bunch of Django packages for integrating with various card processors though.\\

**Betty response:**

Yes.  the ticket you are pointing to is done....  that's how our Act bid/Vendor bid functions now.

I also wrote:

https://github.com/bethlakshmi/gbe-divio-djangocms-python2.7/issues/136

But that was just for a better/modern Paypal.

1. Well that sounds easy... but if BPT has gotten horrible branding from being crooks, will we have problems with redirecting people to them and them not trusting the platform?  My favorite from a tech POV, but not if it hurts the business in any way.

2. To do that, I would want to estimate the integration work based on what system we want.  I think we want:
- for users to be able to easily buy stuff online
- for the front desk to be able to easily sell stuff in person - like with Square
- for the front desk to be able to easily check people in to give out tickets/badges and all.  We don't really need it to manage the physical thing we give (the way we got printed BPT tickets), we can make that stuff up ourselves.  But we want to be able to check a box when someone takes their stuff, so they can't take it a second time later.  Preferably with a separate login so we aren't all super-user.
- keeps people from buying stuff if we're sold out - particularly for complicated scenarios, like giving away tickets as part of a package as well as selling them as individual units.
- handles money in compliance with the law

Sounds like Simple Tix provides at least some if not all of those.  If so I'm glad to move forward on estimating that.

Looking at SimpleTix, specifically, it looks like they have an API integration option.  It integrates through Zapier (new one on me...), which seems delightfully (deceptively?) simple.  There is also a way to email ticket buyers through Simple Tix via Zapier.  And I think we *might* be able to do other cool stuff.  It does a LOT of different very commonly used service/platform integrations.  A whole bunch of it is to let people who are more like you and less like me, plug stuff together without having to deal with people like me. :)  

The downside of this integration with this particular combo of apps is that Zapier may have a pricing model we don't like. If I understand it, the free version is 100 tasks a month.  That would mean we get capped (I think) at having Simple Tix send us 100 messages a month - where every ticket purchase could be at least 1 message.  The $15/month plan would work (750 messages).  And I can't tell if there are other limits around me doing what I would need to do.  I could get into it with a customer support person...

The integration points look workable.  They will send us a notification when:
- a new attendee connects to us through SimpleTix
- a new order gets purchased at SimpleTix

The BPT method was a polling request.  This is a push.  That's different but totally fine.  I think I could adapt it pretty well on a data storage side, so that I'm mostly mushing their notifications into our structures.

3.  Yep.  That's more work.  I don't know payment systems as well as you probably do - but Square seems like the way to go - it has both online and on-premise payment methods.  It's widely used.  There is at least a bit of a product catalog (says the docs), which might help a bit.  But all the bullet points in item #2 just became coding work for me in item #3.  And chances are good, that my GUI will not be as nice as SimpleTix, since they clearly have a team of well paid staff working on this, including people with skills I don't have that make web pages easier to use.

My biggest fear on #3 is the complexity of marketing and coupons that you use to promote the side.  I can greatly simplify giving comps and discounts to performers, since it's all in 1 DB.  But you have at times had coupon codes, timed ticket prices (x price until date, then y price), limited ticket counts (first 10 tickets are X, then they are Y $), and combos of all that... getting all the logic you want and at the time that you want it will be a challenge.  I strongly suspect we may have some frustrating conversations where I didn't understand the feature you want and you need it NOW for business promotional reasons.

SUMMARY:

- 1 BPT - lack of trust, also not super easy to use (last I tried) as the customer.
- 2 SimpleTix - could be expensive (Zapier + SimpleTix), likely to be nicer to use, integration is not too hard.  Would recommend it for ticket buying.  Reason NOT to use would be if it doesn't have all the features you need as a ticket seller.
- 3 DIY w. Square - a lot more work, but doable.  Downside - not as pretty & easy to use as pay-for service.  No publicity/brand recognition.  Upside - we get exactly the business logic we want.... eventually...

**Comment by bethlakshmi** [comment]
I thought this was in ticket form - but it was an email.  Putting it in as a ticket so we can continue.  Looks like I did a serious review of Simple Tix, but not much on PayPal.  WIll look there.

**Comment by bethlakshmi** [comment]
OK, I got my arms around setting up a fake account and doing a really primitive integration - from SimpleTix to Google Sheets (almost).  I stopped short, because the same data for "Zap" that reflects an order in SimpleTix lacks much concrete information about what they purchased.  The "Ticket Type" looks like this - "test 6/30/2021 8:00 PM Section: testy"
- Test is the name of the event (so each event in GBE would have to be uniquely named)
- 6/30/2021 8:00 PM - date and time of the event - OK fine.
- Section: testy - the type of ticket they bought.
I can also get price information - both what it costs and what they paid - both per item and total transaction.

Also, there is no "Zap" for a ticket or price being created.  The other two zaps are - new attendee created, new membership created.  So all sync between the system is customer triggered.  As far as I can, there is no way of pushing pure ticket data to the system.

Challenges on the GBE side of the house:
- we have a whole flow about syncing tickets - we sync tickets and prices, connect them to stuff on our site, and then publish price info.  I don't see a way to do that here.
- linking what a user bought to the actual event is doable, but brittle.  There's no "ID" here - the event name is about all I have to work with - so I need to make sure that the simple tix event name and the GBE event name are always the same, or devise a secondary "simple tix event name" to connect them.  Since the event name is also user visible in simple tix, it can't be too goofy.

Side note - I was non-plussed by how slow Simple tix is on the user end.  I made 1 sample event.  And 1 type of ticket.  On 1 day.  I clicked on the link they gave me for users, and waited for *minutes*.  After the long wait, I had to pick a date (why??  it happens on 1 day!?) - and had a second several minute long wait.  That's not good - will users live through it?

**Comment by bethlakshmi** [comment]
OK, working on eventbrite - making good process.

Assembling questions and decisions here:

QUESTIONS (To @burlexpo ):
1 - Do I rip out BPT completely?  I have these options:

- Rip it all out.  No more code, hard to put back but not impossible.  Benefit - eliminates complexity and test time.
- Make it not-functional, but leave a lot of code in place.  It would take work to make it functional again, but less so than the first option. 
- Make it possible to have the system do either one - put in a switch to say what we're doing right now. 
- Make it possible for the system to do both at once.

Options 1 & 2 are equally easy.  Options 3 & 4 are more work and a bit of complexity for the user as the BPT event create flow is by necessity different than the event brite event create flow (which is easier, because of the Organization concept).

2 - Do we keep GBE 2020 ticket purchases that were made in BPT?  Were all those refunded and should be removed?  Or are some still valid?  I can:
- remove selected types (ie, remove all the whole shebangs, but leave the vendors) or manually pull out tickets that were refunded and leave the rest
- leave them all - no additional work - porting the data is already done.
- remove them all - trivial work, delete is easy


Decisions:
- I am keeping the polling model, in the interest of speed and sanity.  There was an alternative option - using webhooks - which would work on a push model.  When something of interest to us happens in Eventbrite, we can tell it to tell us.  That's actually very nice, as we find out immediately, and we avoid polling and polling when nothing interesting is happening.  BUT - it's a serious pattern change, and I feared the extra effort, given that we have a timeline here.
- I will be syncing ALL data - events, tickets, purchases - periodically.  We used on only sync purchases - and that sometimes created bad surprises.  New Events will not be properly linked to anything in GBE without manual effort - but they WILL show up on the ticketing page and in a user's profile as something they bought.
- the "Sync" button on the ticket management page will give a bunch more error communication than it used to - this was always my go-to for "what went wrong?" and so I beefed it up a bunch with messaging that comes back to the page when the sync fails.  I will likely do the same with sync transactions.

**Comment by bethlakshmi** [comment]
More questions - further into the integration:
- Eventbrite has a cost, a fee, tax, and a few other less clear fields for each ticket.  We have a cost field.  Which is what we show to users (really we show the min and the max of the visible ticket costs, nothing else).   
- Eventbrite also has an "include_fee" field based on whether we've chosen for the fee to be added to the cost, or absorbed into the cost.  Ie, if include_fee is False - if the ticket is $100 and the fee is $7, the user is charged $107.  If it's True - the charge to the user is $100.  

.... I'm unsure how taxes fit in.

Question - I'm assuming I should use that setting and calculate what the actual charge would be on our site.  

Alternative - just use cost.  Figure that people are used to fees?

**Comment by burlexpo** [comment]
I think just use cost.  That way people can be annoyed with EventBrite and not us -- and it makes it clear that the money is going to the ticket vendor.

**Comment by bethlakshmi** [comment]
Cool.  Done.  See the post above this with questions about how much to rip out BPT.

**Comment by burlexpo** [comment]
I think for now just leave the old BPT ticketing stuff in place unless it’s going to interfere with something. 

--
Team BurlExo
Burlesque-Expo.com
#BurlExpo

**Comment by bethlakshmi** [comment]
I have managed to keep it functional.  On the manual import side (when you push a button and it imports events/tickets or transactions) - the system is noticeably slower because it tries both ways and there's not a clear way right now to have an event labeled for BPT vs. EB (eventbrite).  I could do that, but that's work.

On the automatic side, I'm gonna keep sanity by having separate sync scripts for BPT vs. EB.  I recommend that if we aren't using BPT, we turn those off, so the system isn't working hard for no good reason.

**Comment by bethlakshmi** [comment]
OK, new progress - and questions for @burlexpo 

Progress:
- events, tickets, and transactions are all getting imported manually - which is more than 50% of it.
- tested event upload pagination - fixing & testing now - need to do the same for tickets and for transactions.

Questions/issues:

**BPT really slowing things down.**

Not in terms of development speed - in terms of computational speed.  A big thing is that I don't have a way of marking BPT events vs. Event Brite events - we have a "source" on tickets, but not events.  That means that both systems are trying to sync all events.

I also started getting an SSL certificate validity failure in my local environment when calling BPT that I'm not terribly interested in debugging right now as there's no real value to it right now.

Proposal 1 - put in a switch for "are we using BPT right now?" - if no, stop all syncing with it.  That stops the BPT calls (and my problems with them) - but there is still some inefficiency on ticket syncing with Eventbrite due to events that are not in EB.

Proposal 2 - put in an event level switch for BPT vs. Eventbrite vs. Paypal for the event type level - that keeps me from calling EB or BPT when the call is irrelevant - but it doesn't remove the times when we call BPT knowing there will be no change (and so the SSL cert errors will still be a thing

Proposal 3 - do both.  Best of both worlds, full flexibility/efficiency - but a bit more work for me.

**Event publication state**

The EB events & tickets have several states.  Boiling it down to what I really care about:
- Events can be published or in a draft state - draft events are not visible to anyone. 
- Tickets can be visible or hidden.  An event cannot be published if it has no visible tickets.

It's easy enough to mimic ticket state in our system - we have the same basic idea.  The harder part is event state.  We don't have a draft/published state.  An event is live if it has visible tickets.  

My current inclination is to not try to absorb the draft/published event state.  We don't update events/tickets after the first sync, because it's impossible to figure out what the user wants to keep/change - but we do update any new tickets within an event.  So I'm going to assume that if a Draft Event in Event Brite is synced with GBE and it gets a visible ticket, then it's OK for GBE to show that visible ticket (and the event).  

The alternative would be to not even try to sync an event until its published - which means it's live on EB before it's visible in GBE.

Any preference?  I don't feel strongly I'm just picking something.

**Comment by bethlakshmi** [comment]
I've got this integrated on the test server, if you want to play with it.  It's using my account with a bunch of crappy test data.  If you want to play with it in your own account, I need an Oauth token and org id.

We should book time soon to set up the real account you want to - probably worth it to do a dry run in test first.

**Comment by bethlakshmi** [comment]
OK - it's live and configured!

It imported 3 events and a bunch of tickets.  Everything looks good.
I took the liberty of manually hooking in how the tickets work on our site - ie, that the Whole Shebang covers the conference all tickets, and with show tickets go with which scheduled show.

Due to the migration, this public facing page:

https://www.burlesqueexpo.com/ticketing/

Looked kind of strange, because the old BPT versions of events are still "live" - so I will make sure that the BPT versions are hidden.

Also - I realize I left a gap - on that public page, the purchase link is a redirect to BPT!!  But for Event Brite it should go to EB - so working on that very quickly now.

**Comment by bethlakshmi** [comment]
OK - I fixed:
https://www.burlesqueexpo.com/ticketing/
and event pages like this:
https://www.burlesqueexpo.com/scheduling/details/1084

which have an automatically made link.
-----------
Also - I verified that my auto-syncer is happily running at 15 and 45 past the hour (every 1/2 hour), and reporting it's status in the sync status table I just built.

Will leave this open for a few days while get the rest of the tickets set up in EB.
